utils/trie_writer.hh

```diff
@@ -0,0 +1,1124 @@
+        std::ranges::copy(blob, _payload_buf.data());
+    }
+    bool operator==(const trie_payload& other) const {
+        return other._payload_bits == _payload_bits
+            && std::ranges::equal(other.blob(), blob());
+    }
+
+    const_bytes blob() const noexcept{
+        return {_payload_buf.data(), _payload_size};
+    }
+};
+
+// The below few types are just stronly-typed wrappers around integers,
+// just so they can have some extra correctness checks and
+// so that we can write .valid() instead of "!= -1".
+// I'm not sure if this is better or worse than just using regular integers.
+
+// Represents a difference between positions in the output stream.
+// Just a strongly-typed wrapper around int64_t.
+struct sink_offset {
+    int64_t value = -1;
+    sink_offset() = default;
+    explicit sink_offset(int64_t v) : value(v) {}
+    std::strong_ordering operator<=>(const sink_offset& o) const {
+        EXPENSIVE_ASSERT(valid());
+        EXPENSIVE_ASSERT(o.valid());
+        return value <=> o.value;
+    };
+    bool operator==(const sink_offset& o) const = default;
+    bool valid() const {
+        return value >= 0;
+    }
+    sink_offset operator+(sink_offset o) const {
+        EXPENSIVE_ASSERT(valid());
+        EXPENSIVE_ASSERT(o.valid());
+        return sink_offset{value + o.value};
+    }
+};
+
+// Represents a size of a serialized node.
+// Just a strongly-typed wrapper around int16_t.
+struct node_size {
+    int16_t value = -1;
+    node_size() = default;
+    explicit node_size(int16_t v) : value(v) {}
+    bool valid() const {
+        return value >= 0;
+    }
+    operator sink_offset() const {
+        return sink_offset(value);
+    }
+};
+
+// Represents a position in the output stream.
+// Just a strongly-typed wrapper around int64_t.
+struct sink_pos {
+    int64_t value = -1;
+    sink_pos() = default;
+    explicit sink_pos(int64_t v) : value(v) {}
+    bool valid() const {
+        return value >= 0;
+    }
+    std::strong_ordering operator<=>(const sink_pos& o) const {
+        EXPENSIVE_ASSERT(valid());
+        EXPENSIVE_ASSERT(o.valid());
+        return value <=> o.value;
+    };
+    sink_pos operator+(sink_offset o) const {
+        EXPENSIVE_ASSERT(valid());
+        EXPENSIVE_ASSERT(o.valid());
+        return sink_pos{value + o.value};
+    }
+    sink_pos operator-(sink_offset o) const {
+        EXPENSIVE_ASSERT(valid());
+        EXPENSIVE_ASSERT(o.valid());
+        return sink_pos{value - o.value};
+    }
+    sink_offset operator-(sink_pos pos) const {
+        EXPENSIVE_ASSERT(valid());
+        EXPENSIVE_ASSERT(pos.valid());
+        return sink_offset{value - pos.value};
+    }
+};
+
+class writer_node;
+
+// trie_writer consumes a stream of keys and produces a stream of writer_node objects,
+// which is fed into a trie_writer_sink.
+template <typename T>
+concept trie_writer_sink = requires(T& o, const T& co, const writer_node& x, sink_pos pos) {
```

Comment:
What's the motivation for making trie_writer a template?

Testability would be a good answer.